pkg/autodiscovery/providers/prometheus_pods.go

```diff
@@ -0,0 +1,151 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build kubelet
+
+package providers
+
+import (
+	"encoding/json"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/providers/names"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/kubelet"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// PrometheusPodsConfigProvider implements the ConfigProvider interface for prometheus pods.
+type PrometheusPodsConfigProvider struct {
+	kubelet kubelet.KubeUtilInterface
+	checks  []*PrometheusCheck
+}
+
+// NewPrometheusPodsConfigProvider returns a new Prometheus ConfigProvider connected to kubelet.
+// Connectivity is not checked at this stage to allow for retries, Collect will do it.
+func NewPrometheusPodsConfigProvider(config config.ConfigurationProviders) (ConfigProvider, error) {
+	p := &PrometheusPodsConfigProvider{}
+	err := p.setupConfigs()
+	return p, err
+}
+
+// setupConfigs reads and initializes the checks from the configuration
+// It defines a default openmetrics instances with default AD if the checks configuration is empty
+func (p *PrometheusPodsConfigProvider) setupConfigs() error {
+	checks := []*PrometheusCheck{}
+	if err := config.Datadog.UnmarshalKey("prometheus_scrape.checks", &checks); err != nil {
+		return err
+	}
+
+	if len(checks) == 0 {
+		log.Info("The 'prometheus_scrape.checks' configuration is empty, a default openmetrics check configuration will be used")
+		p.checks = []*PrometheusCheck{defaultCheck}
+		return nil
+	}
+
+	for i, check := range checks {
+		if err := check.init(); err != nil {
+			log.Errorf("Ignoring check configuration (# %d): %v", i+1, err)
+			continue
+		}
+		p.checks = append(p.checks, check)
+	}
+
+	return nil
+}
+
+// String returns a string representation of the PrometheusPodsConfigProvider
+func (p *PrometheusPodsConfigProvider) String() string {
+	return names.Prometheus
+}
+
+// Collect retrieves templates from the kubelet's podlist, builds config objects and returns them
+func (p *PrometheusPodsConfigProvider) Collect() ([]integration.Config, error) {
+	var err error
+	if p.kubelet == nil {
+		p.kubelet, err = kubelet.GetKubeUtil()
```

Comment:
This initialization doesn't need to be threadsafe, correct?

Reply:
no, only one main AD goroutine calls `Collect`

Reply:
(this is exactly how it's initialized in the kubelet config provider as well)